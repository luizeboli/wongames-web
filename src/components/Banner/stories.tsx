import { Meta, Story } from '@storybook/react/types-6-0';

import Banner, { BannerProps } from '.';

export default {
  title: 'Banner',
  component: Banner,
  parameters: {
    layout: 'fullscreen',
  },
} as Meta;

export const Basic: Story<BannerProps> = (args) => <Banner {...args} />;

Basic.args = {
  img: 'https://source.unsplash.com/user/willianjusten/1042x580',
  title: 'In the Game',
  subtitle: '<p>Your best choice</p>',
  buttonLabel: 'Buy',
  buttonLink: '#',
};

export const withRibbon: Story<BannerProps> = (args) => (
  <div style={{ maxWidth: '54rem', margin: '0 auto' }}>
    <Banner {...args} />
  </div>
);

withRibbon.args = {
  ribbon: 'Hot Deal',
  ribbonSize: 'normal',
  ribbonColor: 'secondary',
};
