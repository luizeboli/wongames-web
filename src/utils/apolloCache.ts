import { InMemoryCache } from '@apollo/client/cache';

export default new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        games: {
          keyArgs: ['filters', 'sort'],
          merge(existing, incoming) {
            if (!existing) return incoming;

            return {
              data: [...existing.data, ...incoming.data],
              meta: incoming.meta,
            };
          },
        },
      },
    },
    Wishlist: {
      fields: {
        games: {
          merge(_, incoming) {
            return incoming;
          },
        },
      },
    },
  },
});
