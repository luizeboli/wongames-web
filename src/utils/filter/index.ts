import { ParsedUrlQueryInput } from 'querystring';

import { ItemProps } from 'components/ExploreSidebar';

type ParseArgs = {
  queryString: ParsedUrlQueryInput;
  filterItems?: Pick<ItemProps, 'type' | 'name'>[];
};

/*
 * Receives a query string and parse it to GraphQL where format
 */
export const parseQueryStringToWhere = ({ queryString, filterItems }: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {};

  Object.keys(queryString)
    .filter((item) => item !== 'sort')
    .forEach((key) => {
      const item = filterItems?.find((item) => item.name === key);
      const isCheckbox = item?.type === 'checkbox';

      obj[key] = !isCheckbox ? queryString[key] : { name_contains: queryString[key] };
    });

  return obj;
};

/*
 * Receives a query string and parse it to filter format to use in explore page
 */
export const parseQueryStringToFilter = ({ queryString, filterItems }: ParseArgs) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const obj: any = {};

  Object.keys(queryString).forEach((key) => {
    const item = filterItems?.find((item) => item.name === key);
    const isCheckbox = item?.type === 'checkbox';
    const isArray = Array.isArray(queryString[key]);

    obj[key] = !isArray && isCheckbox ? [queryString[key]] : queryString[key];
  });

  return obj;
};
