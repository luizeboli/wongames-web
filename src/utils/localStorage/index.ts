const APP_KEY = 'WONGAMES';

export const getStorageItem = (key: string) => {
  if (typeof window === 'undefined') return;

  const data = window.localStorage.getItem(`${APP_KEY}_${key}`);

  return JSON.parse(data!);
};

export const setStorageItem = (key: string, value: string[]) => {
  if (typeof window === 'undefined') return;

  window.localStorage.setItem(`${APP_KEY}_${key}`, JSON.stringify(value));
};
