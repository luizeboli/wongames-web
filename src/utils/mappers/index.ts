import { TGamesFragment, THighlightFragment, TQueryHome } from 'graphql/generated';
import formatPrice from 'utils/formatPrice';

export const bannersMapper = (banners: TQueryHome['banners']) =>
  banners?.data.map((banner) => ({
    img: banner.attributes?.image.data?.attributes?.url,
    title: banner.attributes?.title,
    subtitle: banner.attributes?.subtitle,
    buttonLabel: banner.attributes?.button?.label,
    buttonLink: banner.attributes?.button?.link,
    ...(banner.attributes?.ribbon && {
      ribbon: banner.attributes.ribbon.text,
      ribbonColor: banner.attributes.ribbon.color,
      ribbonSize: banner.attributes.ribbon.size,
    }),
  }));

export const gamesMapper = (games: TGamesFragment | null | undefined) =>
  games?.data.length
    ? games.data.map((game) => ({
        id: game.id,
        title: game.attributes?.name,
        slug: game.attributes?.slug,
        cover: game.attributes?.cover?.data?.attributes?.url,
        ...(!!game.attributes?.developers?.data.length && {
          developer: game.attributes.developers.data[0].attributes?.name,
        }),
        img: game.attributes?.cover?.data?.attributes?.url,
        price: game.attributes?.price,
      }))
    : [];

export const highlightMapper = (hightlight: THighlightFragment | null | undefined) =>
  hightlight
    ? {
        title: hightlight.title,
        subtitle: hightlight.subtitle,
        backgroundImage: hightlight.background?.data?.attributes?.url,
        floatImage: hightlight.floatImage?.data?.attributes?.url ?? '',
        buttonLabel: hightlight.buttonLabel,
        buttonLink: hightlight.buttonLink,
        alignment: hightlight.alignment,
      }
    : null;

export const cartMapper = (games: QueryGames_games[] | QueryWishlist_wishlists_games[] | undefined) => {
  return games
    ? games.map((game) => ({
        id: game.id,
        img: game.cover?.url,
        title: game.name,
        price: formatPrice(game.price),
      }))
    : [];
};

export const ordersMapper = (orders: QueryOrders_orders[]) => {
  return orders
    ? orders.map((order) => {
        return {
          id: order.id,
          paymentInfo: {
            flag: order.card_brand,
            img: order.card_brand ? `/img/cards/${order.card_brand}.png` : null,
            number: order.card_last4 ? `**** **** **** ${order.card_last4}` : 'Free Game',
            purchaseDate: `Purchase made on ${new Intl.DateTimeFormat('en-US', {
              day: 'numeric',
              month: 'short',
              year: 'numeric',
            }).format(new Date(order.created_at))}`,
          },
          games: order.games.map((game) => ({
            id: game.id,
            title: game.name,
            downloadLink: 'https://wongames.com/game/download/yuYT56Tgh431LkjhNBgdf',
            img: game.cover?.url,
            price: formatPrice(game.price),
          })),
        };
      })
    : [];
};
