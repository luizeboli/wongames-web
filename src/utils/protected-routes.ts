import { GetServerSidePropsContext } from 'next';
import { getSession } from 'next-auth/client';

async function protectedRoutes(context: GetServerSidePropsContext) {
  const session = await getSession(context);

  console.log({ session }, 'session');

  if (!session) {
    context.res.writeHead(302, {
      Location: '/sign-in',
    });

    context.res.end();
  }

  return session;
}

export default protectedRoutes;
