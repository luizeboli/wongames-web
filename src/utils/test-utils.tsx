import { ReactElement } from 'react';
import { ThemeProvider } from 'styled-components';
import { render, RenderOptions } from '@testing-library/react';

import { CartContext, CartContextData, CartContextDefaultValues } from 'hooks/use-cart';
import { WishlistContext, WishlistContextData, WishlistContextDefaultValues } from 'hooks/use-wishlist';
import theme from 'styles/theme';

type CustomRenderProps = {
  cartProviderProps?: CartContextData;
  wishlistProviderProps?: WishlistContextData;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <WishlistContext.Provider value={wishlistProviderProps}>
        <CartContext.Provider value={cartProviderProps}>{ui}</CartContext.Provider>
      </WishlistContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
