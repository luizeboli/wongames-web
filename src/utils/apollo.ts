import { useMemo } from 'react';
import { Session } from 'next-auth';
import { ApolloClient, from, HttpLink, NormalizedCacheObject } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';
import { onError } from '@apollo/client/link/error';

import apolloCache from './apolloCache';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ message, locations, path }) =>
      console.log(`[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`),
    );
  if (networkError) console.log(`[Network error]: ${networkError}`);
});

const httpLink = new HttpLink({ uri: process.env.NEXT_PUBLIC_GRAPHQL_API });

function createApolloClient(session?: Session | null) {
  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || '';
    const Authorization = jwt ? `Bearer ${jwt}` : '';

    return { headers: { ...headers, Authorization } };
  });

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: from([authLink, errorLink, httpLink]),
    cache: apolloCache,
  });
}

export function initializeApollo(initialState = null, session?: Session | null) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session);

  if (initialState) {
    apolloClientGlobal.cache.restore(initialState);
  }

  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') return apolloClientGlobal;
  // Create the Apollo Client once in the client
  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = null, session?: Session | null) {
  const store = useMemo(() => initializeApollo(initialState, session), [initialState, session]);
  return store;
}
