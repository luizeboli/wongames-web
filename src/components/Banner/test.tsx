import { render, screen } from 'utils/test-utils';

import Banner from '.';

const props = {
  img: 'image-source',
  title: 'Game Title',
  subtitle: 'Game Subtitle',
  buttonLabel: 'Buy Now',
  buttonLink: 'buy-game',
};

describe('<Banner />', () => {
  it('should render correctly', () => {
    const { container } = render(<Banner {...props} />);

    expect(screen.getByRole('heading', { name: /Game Title/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Game Subtitle/i })).toBeInTheDocument();
    expect(screen.getByRole('image')).toHaveAttribute('src', 'image-source');
    expect(screen.getByRole('link', { name: /Buy Now/i })).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a Ribbon', () => {
    render(<Banner {...props} ribbon="Hot Deal" ribbonSize="small" ribbonColor="secondary" />);

    const ribbon = screen.getByText(/Hot Deal/i);

    expect(ribbon).toBeInTheDocument();
    expect(ribbon).toHaveStyle({
      backgroundColor: '#3cd3c1',
    });
    expect(ribbon).toHaveStyle({
      height: '2.6rem',
      fontSize: '1.2rem',
    });
  });
});
