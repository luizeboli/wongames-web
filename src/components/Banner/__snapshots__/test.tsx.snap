// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
.c5 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #FAFAFA;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  border: none;
  border-radius: 0.4rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

.c5:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

.c0 {
  position: relative;
}

.c1 {
  width: 100%;
  height: 23rem;
  background-color: #EAEAEA;
  background-image: url(image-source);
  background-position: center center;
  background-size: cover;
}

.c2 {
  width: 100%;
  background-color: rgba(0,0,0,0.7);
  padding: 2.4rem;
}

.c3 {
  font-size: 1.8rem;
  font-weight: 600;
  color: #FAFAFA;
}

.c4 {
  color: #FAFAFA;
  font-size: 1.4rem;
  font-weight: 400;
  margin-bottom: 1.6rem;
}

.c4 strong {
  color: #F231A5;
  font-weight: 600;
}

@media (max-width:1170px) {
  .c0 .sc-gtssRu {
    right: 0;
  }

  .c0 .sc-gtssRu::before {
    display: none;
  }
}

@media (min-width:768px) {
  .c0 {
    box-shadow: 0 0.4rem 0 0.5rem 0 rgba(0,0,0,0.2);
  }
}

@media (min-width:768px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width:768px) {
  .c2 {
    position: absolute;
    bottom: 0;
    left: 0;
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 4.0rem;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 2.8rem;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 1.8rem;
  }
}

<div
  class="c0"
>
  <div
    aria-label="Game Title"
    class="c1"
    role="image"
    src="image-source"
  />
  <div
    class="c2"
  >
    <h2
      class="c3"
    >
      Game Title
    </h2>
    <h3
      class="c4"
    >
      Game Subtitle
    </h3>
    <a
      class="c5"
      href="buy-game"
    >
      <span>
        Buy Now
      </span>
    </a>
  </div>
</div>
`;
